### PR TITLE
[SD-323] deployment metadata

### PR DESCRIPTION
**Pull request by GROwen**
### Motivation
Addresses dpc-sdp/bay-cli#25

### Usage
```
> bay deployment metadata
{"deployment":{"sha":"58d27fb5218e4703bd7c1a696471f69f259226cd","authorName":"Guy Owen","when":"2024-11-25 12:36:16 +1100 +1100","tag":"tag not found","msg":"[SD-323] Added deployment cmd."}}
```

### Notes
~I think there's a number of refinements that could be made that I'll point out in PR comments~ Done

[SD-323]: https://digital-vic.atlassian.net/browse/SD-323?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cmd/deployment/metadata.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	errors "github.com/pkg/errors"
+	"github.com/urfave/cli/v2"
+)
+
+type Metadata struct {
+	Deployment Item `json:"deployment"`
+}
+
+type Item struct {
+	Sha        string `json:"sha"`
```
Comment (GROwen) [inline cmd/deployment/metadata.go:18]: I struggled with simplifying this into a single custom struct `Deployment` as a `map[string]string` and compiling the string in a 'clean' way. This approach made more sense/was cleaner _but_ I've got a hunch it's inefficient/overly verbose?

File: cmd/deployment/metadata.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+		return errors.Wrap(err, "unable to get HEAD reference")
+	}
+	fmt.Println(ref.Hash())
+
+	msg, err := repo.CommitObject(ref.Hash())
+	if err != nil {
+		return errors.Wrap(err, "unable to get commit object")
+	}
+
+	tagString := ""
```
Comment (GROwen) [inline cmd/deployment/metadata.go:41]: if no tag is present that value of `tag` on [line 44](https://github.com/dpc-sdp/bay-cli/pull/27/files#diff-1262dd46af24983711b1ba1a2507ed3ab257004339bbcddfdc43871071ce572eR44) is `<nil>` which I couldn't output as a string. This handling to get a string works but is it brittle or obsfucate errors by hijacking the err on [line 47](https://github.com/dpc-sdp/bay-cli/pull/27/files#diff-1262dd46af24983711b1ba1a2507ed3ab257004339bbcddfdc43871071ce572eR47)?

File: cmd/deployment/metadata.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+	item := Item{
+		Sha:        ref.Hash().String(),
+		AuthorName: msg.Author.Name,
+		When:       msg.Author.When.String(),
+		Msg:        msgFirstLn,
+		Tag:        tagString,
+	}
+
+	md := Metadata{
```
Comment (GROwen) [inline cmd/deployment/metadata.go:61]: Related to my first comment about being overly verbose, this feels unnecessary if the Deployment struct was set up differently.

File: main.go [hunk 4]
```diff
@@ -92,6 +93,18 @@ func main() {
 					},
 				},
 			},
+			{
+				Name:  "deployment",
+				Usage: "commands for deployment actions",
+				Subcommands: []*cli.Command{
+					{
+						Name:      "metadata",
+						Usage:     "generates a file with deployment metadata",
```
Comment (nicksantamaria) [inline main.go:102]:
```suggestion
						Usage:     "generates a json object with deployment metadata",
```

File: cmd/deployment/metadata.go (hunk 1 above)
Comment (nicksantamaria) [reply cmd/deployment/metadata.go:18]: Nope this is pretty much perfect. `map[string]string` works best for single-level objects with dynamic attributes. In this case we know what data we want to capture from git and render into the metadata object.

File: cmd/deployment/metadata.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	"github.com/go-git/go-git/v5"
+	errors "github.com/pkg/errors"
+	"github.com/urfave/cli/v2"
+)
+
+type Metadata struct {
+	Deployment Item `json:"deployment"`
+}
+
+type Item struct {
```
Comment (nicksantamaria) [inline cmd/deployment/metadata.go:17]: I suggestion refactoring this struct name to `Deployment` rather than `Item`

**Review by nicksantamaria — CHANGES_REQUESTED** [review]

**Review by nicksantamaria — APPROVED** [review]

**Pull request merged**